subsys/shell/shell_telnet.c

```diff
@@ -0,0 +1,493 @@
+			  int error,
+			  void *user_data)
+{
+	if (error) {
+		LOG_ERR("Error %d", error);
+		goto error;
+	}
+
+	if (sh_telnet->client_ctx) {
+		LOG_WRN("A telnet client is already in.");
+		goto error;
+	}
+
+	if (net_context_recv(client, telnet_recv, 0, NULL)) {
+		LOG_ERR("Unable to setup reception (family %u)",
+			net_context_get_family(client));
+		goto error;
+	}
+
+	LOG_DBG("Telnet client connected (family AF_INET%s)",
+		net_context_get_family(client) == AF_INET ? "" : "6");
+
+	sh_telnet->client_ctx = client;
+
+	return;
+error:
+	net_context_put(client);
+}
+
+static void telnet_setup_server(struct net_context **ctx, sa_family_t family,
+				struct sockaddr *addr, socklen_t addrlen)
+{
+	if (net_context_get(family, SOCK_STREAM, IPPROTO_TCP, ctx)) {
+		LOG_ERR("No context available");
+		goto error;
+	}
+
+	if (net_context_bind(*ctx, addr, addrlen)) {
+		LOG_ERR("Cannot bind on family AF_INET%s",
+			family == AF_INET ? "" : "6");
+		goto error;
+	}
+
+	if (net_context_listen(*ctx, 0)) {
+		LOG_ERR("Cannot listen on");
+		goto error;
+	}
+
+	if (net_context_accept(*ctx, telnet_accept, K_NO_WAIT, NULL)) {
+		LOG_ERR("Cannot accept");
+		goto error;
+	}
+
+	LOG_DBG("Telnet console enabled on AF_INET%s",
+		family == AF_INET ? "" : "6");
+
+	return;
+error:
+	LOG_ERR("Unable to start telnet on AF_INET%s",
+		family == AF_INET ? "" : "6");
+
+	if (*ctx) {
+		net_context_put(*ctx);
+		*ctx = NULL;
+	}
+}
+
+static int telnet_init(void)
+{
+#ifdef CONFIG_NET_IPV4
+	struct sockaddr_in any_addr4 = {
+		.sin_family = AF_INET,
+		.sin_port = htons(TELNET_PORT),
+		.sin_addr = INADDR_ANY_INIT
+	};
+	static struct net_context *ctx4;
+#endif
+#ifdef CONFIG_NET_IPV6
+	struct sockaddr_in6 any_addr6 = {
+		.sin6_family = AF_INET6,
+		.sin6_port = htons(TELNET_PORT),
+		.sin6_addr = IN6ADDR_ANY_INIT
+	};
+	static struct net_context *ctx6;
+#endif
+
+#ifdef CONFIG_NET_IPV4
+	telnet_setup_server(&ctx4, AF_INET,
+			    (struct sockaddr *)&any_addr4,
+			    sizeof(any_addr4));
+#endif
+#ifdef CONFIG_NET_IPV6
+	telnet_setup_server(&ctx6, AF_INET6,
+			    (struct sockaddr *)&any_addr6,
+			    sizeof(any_addr6));
+#endif
+
+	LOG_INF("Telnet shell backend initialized");
+
+	return 0;
+}
+
+/* Shell API */
+
+static int init(const struct shell_transport *transport,
+		const void *config,
+		shell_transport_handler_t evt_handler,
+		void *context)
+{
+	int err;
+
+	sh_telnet = (struct shell_telnet *)transport->ctx;
+
+	err = telnet_init();
+	if (err != 0) {
+		return err;
+	}
+
+	memset(sh_telnet, 0, sizeof(struct shell_telnet));
+
+	sh_telnet->shell_handler = evt_handler;
+	sh_telnet->shell_context = context;
+
+	k_fifo_init(&sh_telnet->rx_fifo);
+	k_timer_init(&sh_telnet->send_timer, telnet_send_prematurely, NULL);
+
+	return 0;
+}
+
+static int uninit(const struct shell_transport *transport)
+{
+	if (sh_telnet == NULL) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int enable(const struct shell_transport *transport, bool blocking)
```

Comment:
i see that `blocking` flag is not handled. Does telnet shell works in blocking way? does it require interrupts? if shell enables backend in `blocking` mode, it means that it will be used from any context (even highest priority interrupt). If backend requires context switching then on blocking set to true, backend should be closed.

Please consider that even in on hw platforms it may not be supported, it may be on emulated or native posix.

Reply:
It should not be an issue for TELNET. It does work in a blocking manner by default, but the actual processing of outgoing packets is handled by a separate thread, we do not wait for any IRQs here. In the worst case, we should run out of the network packet buffers and return an error, I don't expect any deadlock here.